README.md
# slider_controller

<a href="https://flutter.dev/"><img src="https://img.shields.io/badge/flutter-website-deepskyblue.svg" alt="Flutter Website"></a>
<a href="https://dart.dev"><img src="https://img.shields.io/badge/dart-website-deepskyblue.svg" alt="Dart Website"></a>
<a href="https://developer.android.com" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-Android-deepskyblue">
</a>
<a href="https://developer.apple.com/ios/" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-iOS-deepskyblue">
</a>
<a href="" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-Web-deepskyblue">
</a>
<a href="" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-Mac-deepskyblue">
</a>
<a href="" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-Linux-deepskyblue">
</a>
<a href="" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-Windows-deepskyblue">
</a>
<a href=""><img src="https://app.codacy.com/project/badge/Grade/dc683c9cc61b499fa7cdbf54e4d9ff35"/></a>
<a href="https://github.com/Mindinventory/slider_controller/blob/master/LICENSE" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/github/license/Mindinventory/slider_controller"></a>
<a href="https://pub.dev/packages/slider_controller"><img src="https://img.shields.io/pub/v/slider_controller?color=as&label=slider_controller&logo=as1&logoColor=blue&style=social"></a>
<a href="https://github.com/Mindinventory/slider_controller"><img src="https://img.shields.io/github/stars/Mindinventory/slider_controller?style=social" alt="MIT License"></a>

A package that gives us a slider controller a custom widget. This widget is used for easy customization for the slider widget.

## Key Features

* Ready to use custom slider widget.
* Easy customization for the slider widget.

# Preview

![slider_controller](https://github.com/Mindinventory/slider_controller/blob/master/assets/slider_controller.gif)

## Basic Usage

Import it to your project file

```dart
import 'package:slider_controller/slider_controller.dart';
```

And add it in its most basic form like it:

```dart
SliderController(
  value: 50.0,
  onChanged: (value) {
    print('slider value : $value');
  },
);
```

### Required parameters of SliderController
------------

| Parameter                      | Description                                                                            |
|--------------------------------|----------------------------------------------------------------------------------------|
| double value                   | Indicates the default slider thumb value between the 0.0 to 100.0                      |
| ValueChanged<double> onChanged | Called during a drag when the user is selecting a new value for the slider by dragging |

### Optional parameters of SliderController
------------

| Parameter                           | Default | Description                                                       |
|-------------------------------------|---------|-------------------------------------------------------------------|
| ValueChanged<double>? onChangeEnd   | -       | Called when the user is done selecting a new value for the slider |
| ValueChanged<double>? onChangeStart | -       | Called when the user starts selecting a new value for the slider  |
| double min                          | 0.0     | Indicates the Minimum value for the slider                        |
| double max                          | 100.0   | Indicates the Maximum value for the slider                        |
| SliderDecoration sliderDecoration   | --      | Used to Decorate the Slider Widget                                |
| bool isDraggable                    | true    | Used to Enable or Disable Drag Gesture of Slider                  |

### Optional parameters of SliderDecoration
------------

| Parameter           | Default         | Description                                                     |
|---------------------|-----------------|-----------------------------------------------------------------|
| Color inactiveColor | Colors.blueGrey | Defines the background color of the inactive part of the slider |
| Color activeColor   | Colors.blue     | Defines the background color of the active part of the slider   |
| Color thumbColor    | Colors.white    | Defines the thumb color of the slider                           |
| double borderRadius | 20.0            | Indicates the rounded border radius for the slider              |
| double height       | 50.0            | Indicates the height of the slider                              |
| bool isThumbVisible | true            | Indicates that slider thumb is visible or not                   |
| double thumbHeight  | 25.0            | Indicates the height of the slider thumb                        |
| double thumbWidth   | 5.0             | Indicates the width of the slider thumb                         |

### Guideline for contributors
------------

* Contribution towards our repository is always welcome, we request contributors to create a pull
  request for development.

### Guideline to report an issue/feature request
------------
It would be great for us if the reporter can share the below things to understand the root cause of
the issue.

* Library version
* Code snippet
* Logs if applicable
* Device specification like (Manufacturer, OS version, etc)
* Screenshot/video with steps to reproduce the issue
* Library used

LICENSE!
------------
**slider_controller**
is [MIT-licensed.](https://github.com/Mindinventory/slider_controller/blob/master/LICENSE)

Let us know!
------------
We’d be really happy if you send us links to your projects where you use our component. Just send an
email to sales@mindinventory.com And do let us know if you have any questions or suggestion
regarding our work.

CHANGELOG.md
## 0.0.5

* Added support for onChangeEnd and onChangeStart callback of SliderController. (thanks to @justbendev)
* Updated package details and code improvements.

## 0.0.4

* Added isDraggable flag for the Drag Gesture of Slider.

## 0.0.3

* Update the dart file formats.

## 0.0.2

* Added the min and max limit for the slider value.

## 0.0.1

* Initial Release.

lib/src/slider_controller.dart
import 'package:flutter/material.dart';

import 'render_slider_controller.dart';
import 'slider_decoration.dart';

///
/// Class to Control the Custom Slider Widget.
///
class SliderController extends LeafRenderObjectWidget {
  const SliderController({
    Key? key,
    required this.value,
    required this.onChanged,
    this.onChangeEnd,
    this.onChangeStart,
    this.min = 0.0,
    this.max = 100.0,
    this.sliderDecoration,
    this.isDraggable = true,
  })  : assert(
          value >= min && value <= max,
          'value should be between the min or 0.0 to max or 100.0',
        ),
        assert(
          min < max,
          'value of the min should be smaller than the max',
        ),
        super(key: key);

  /// Indicates the default slider thumb value
  /// Value is between the min or 0.0 to max or 100.0
  final double value;

  /// Called during a drag when the user is selecting a new value for the slider
  /// by dragging.
  final ValueChanged<double> onChanged;

  /// Called when the user is done selecting a new value for the slider.
  final ValueChanged<double>? onChangeEnd;

  /// Called when the user starts selecting a new value for the slider.
  final ValueChanged<double>? onChangeStart;

  /// Indicates the Minimum value for the slider
  /// If min is null then the default value 0.0 is used
  final double min;

  /// Indicates the Maximum value for the slider
  /// If max is null then the default value 100.0 is used
  final double max;

  /// Used to Decorate the Slider Widget
  /// If decoration is null then the default value of the slider decoration is used
  final SliderDecoration? sliderDecoration;

  /// Used to Enable or Disable Drag Gesture of Slider
  /// Default value true = Enable is used
  final bool isDraggable;

  @override
  RenderObject createRenderObject(BuildContext context) {
    /// Create and return the render object
    return RenderSliderController(
      value: value,
      onChanged: onChanged,
      onChangeEnd: onChangeEnd,
      onChangeStart: onChangeStart,
      min: min,
      max: max,
      sliderDecoration: sliderDecoration ?? SliderDecoration(),
      isDraggable: isDraggable,
    );
  }

  @override
  void updateRenderObject(
    BuildContext context,
    covariant RenderSliderController renderObject,
  ) {
    /// Update the render object
    renderObject
      ..value = value
      ..onChanged = onChanged
      ..onChangeEnd = onChangeEnd
      ..onChangeStart = onChangeStart
      ..min = min
      ..max = max
      ..sliderDecoration = sliderDecoration ?? SliderDecoration()
      ..isDraggable = isDraggable;
  }
}


lib/src/slider_decoration.dart
import 'package:flutter/material.dart';

///
/// Used to Decorate the Slider Widget.
///
class SliderDecoration {
  /// Defines the background color of the inactive part of the slider
  /// Default value [Colors.blueGrey]
  final Color inactiveColor;

  /// Defines the background color of the active part of the slider
  /// Default value [Colors.blue]
  final Color activeColor;

  /// Defines the thumb color of the slider
  /// Default value [Colors.white]
  final Color thumbColor;

  /// Indicates the rounded border radius for the slider
  /// Default value 20.0
  final double borderRadius;

  /// Indicates the height of the slider
  /// Default value [50.0]
  /// The [height] is always bigger then the [thumbHeight]
  /// If not then the slider throw the error
  final double height;

  /// Indicates that slider thumb is visible or not
  /// Default value [true]
  final bool isThumbVisible;

  /// Indicates the height of the slider thumb
  /// Default value [25.0]
  /// The [thumbHeight] is always smaller then the [height]
  /// If not then the slider throw the error
  final double thumbHeight;

  /// Indicates the width of the slider thumb
  /// Default value [5.0]
  final double thumbWidth;

  SliderDecoration({
    this.inactiveColor = Colors.blueGrey,
    this.activeColor = Colors.blue,
    this.thumbColor = Colors.white,
    this.borderRadius = 20.0,
    this.height = 50.0,
    this.isThumbVisible = true,
    this.thumbHeight = 25.0,
    this.thumbWidth = 5.0,
  }) : assert(
          thumbHeight < height,
          'thumbHeight should not be same or bigger than height of slider',
        );
}


lib/src/render_slider_controller.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

import 'slider_decoration.dart';

///
/// Class to Render the Slider Widget.
///
class RenderSliderController extends RenderBox {
  RenderSliderController({
    required double value,
    required ValueChanged<double> onChanged,
    required ValueChanged<double>? onChangeEnd,
    required ValueChanged<double>? onChangeStart,
    required double min,
    required double max,
    required SliderDecoration sliderDecoration,
    required bool isDraggable,
  })  : _value = value,
        _onChanged = onChanged,
        _onChangeEnd = onChangeEnd,
        _onChangeStart = onChangeStart,
        _min = min,
        _max = max,
        _sliderDecoration = sliderDecoration,
        _isDraggable = isDraggable {
    if (_isDraggable) {
      /// Setting up the drag gesture for the slider widget
      _drag = HorizontalDragGestureRecognizer()
        ..onStart = (DragStartDetails details) {
          _updateSliderThumbPosition(
            details.localPosition,
            ChangeValueType.changeStart,
          );
        }
        ..onUpdate = (DragUpdateDetails details) {
          _updateSliderThumbPosition(
            details.localPosition,
            ChangeValueType.change,
          );
        }
        ..onEnd = (DragEndDetails details) {
          _onChangeEnd?.call(_value);
        };
    }
  }

  /// Getter and Setter for the slider widget to render

  /// Indicates the default slider thumb value
  /// Value is between the 0.0 to 100.0
  double get value => _value;
  double _value;

  set value(double val) {
    assert(
      val >= _min && val <= _max,
      'value should be between the min or 0.0 to max or 100.0',
    );
    if (_value == val) {
      return;
    }

    /// Setting the value and calling the paint method to render the slider widget
    _value = val;
    markNeedsPaint();
  }

  /// Called during a drag when the user is selecting a new value for the slider
  /// by dragging.
  ValueChanged<double> get onChanged => _onChanged;
  ValueChanged<double> _onChanged;

  set onChanged(ValueChanged<double> changed) {
    if (_onChanged == changed) {
      return;
    }

    /// Setting the onChanged Function and calling the paint method to render the
    /// slider widget
    _onChanged = changed;
    markNeedsPaint();
  }

  /// Called when the user is done selecting a new value for the slider.
  ValueChanged<double>? get onChangeEnd => _onChangeEnd;
  ValueChanged<double>? _onChangeEnd;

  set onChangeEnd(ValueChanged<double>? changed) {
    if (_onChangeEnd == changed) {
      return;
    }

    /// Setting the onChangeEnd Function and calling the paint method to render the
    /// slider widget
    _onChangeEnd = changed;
    markNeedsPaint();
  }

  /// Called when the user starts selecting a new value for the slider.
  ValueChanged<double>? get onChangeStart => _onChangeStart;
  ValueChanged<double>? _onChangeStart;

  set onChangeStart(ValueChanged<double>? changed) {
    if (_onChangeStart == changed) {
      return;
    }

    /// Setting the onChangeStart Function and calling the paint method to render the
    /// slider widget
    _onChangeStart = changed;
    markNeedsPaint();
  }

  /// Indicates the Minimum value for the slider
  /// If min is null then the default value 0.0 is used
  double get min => _min;
  double _min;

  set min(double val) {
    if (_min == val) {
      return;
    }

    /// Setting the value and calling the paint method to render the slider widget
    _min = val;
    markNeedsPaint();
  }

  /// Indicates the Maximum value for the slider
  /// If max is null then the default value 100.0 is used
  double get max => _max;
  double _max;

  set max(double val) {
    if (_max == val) {
      return;
    }

    /// Setting the value and calling the paint method to render the slider widget
    _max = val;
    markNeedsPaint();
  }

  /// Used to Decorate the Slider Widget
  SliderDecoration get sliderDecoration => _sliderDecoration;
  SliderDecoration _sliderDecoration;

  set sliderDecoration(SliderDecoration decoration) {
    if (_sliderDecoration == decoration) {
      return;
    }

    /// Setting the decoration and calling the paint and layout method to render the
    /// slider widget
    _sliderDecoration = decoration;
    markNeedsPaint();
    markNeedsLayout();
  }

  /// Used to Enable or Disable Drag Gesture of Slider
  bool get isDraggable => _isDraggable;
  bool _isDraggable;

  set isDraggable(bool isDrag) {
    if (_isDraggable == isDrag) {
      return;
    }

    /// Setting the Enable or Disable Behaviour of the Drag Gesture
    _isDraggable = isDrag;
    markNeedsPaint();
    markNeedsLayout();
  }

  /// Default stroke width for the painters
  final double _strokeWidth = 4.0;

  /// Left side padding for the slider thumb
  final double _thumbLeftPadding = 10.0;

  @override
  void performLayout() {
    /// Setting up the size for the slider widget
    final desiredWidth = constraints.maxWidth;
    final desiredHeight = _sliderDecoration.height;
    final desiredSize = Size(desiredWidth, desiredHeight);
    size = constraints.constrain(desiredSize);
  }

  @override
  void paint(PaintingContext context, Offset offset) {
    /// Setting up the canvas to draw the slider widget
    final canvas = context.canvas;
    canvas.save();
    canvas.translate(offset.dx, offset.dy);

    /// Painter for the inactive part of slider
    final inactiveSliderPainter = Paint()
      ..color = _sliderDecoration.inactiveColor
      ..strokeCap = StrokeCap.round
      ..strokeWidth = _strokeWidth;

    /// Painter for the active part of slider
    final activeSliderPainter = Paint()
      ..color = _sliderDecoration.activeColor
      ..strokeCap = StrokeCap.round
      ..strokeWidth = _strokeWidth;

    /// Painter for the slider thumb
    final thumbPainter = Paint()..color = _sliderDecoration.thumbColor;

    /// Drawing the inactive part of slider
    canvas.drawRRect(
      RRect.fromRectAndCorners(
        Offset.zero & Size(constraints.maxWidth, _sliderDecoration.height),
        topRight: Radius.circular(_sliderDecoration.borderRadius),
        topLeft: Radius.circular(_sliderDecoration.borderRadius),
        bottomRight: Radius.circular(_sliderDecoration.borderRadius),
        bottomLeft: Radius.circular(_sliderDecoration.borderRadius),
      ),
      inactiveSliderPainter,
    );

    /// Drawing the active part of slider or bar from left to thumb position
    final thumbDx = ((_value - _min) / (_max - _min)) * size.width;
    canvas.drawRRect(
      RRect.fromRectAndCorners(
        Offset.zero & Size(thumbDx, _sliderDecoration.height),
        topRight: Radius.circular(_sliderDecoration.borderRadius),
        topLeft: Radius.circular(_sliderDecoration.borderRadius),
        bottomRight: Radius.circular(_sliderDecoration.borderRadius),
        bottomLeft: Radius.circular(_sliderDecoration.borderRadius),
      ),
      activeSliderPainter,
    );

    if (_sliderDecoration.isThumbVisible) {
      /// Drawing the slider thumb
      final thumbDesiredDx =
          thumbDx - ((thumbDx == 0.0) ? 0.0 : _thumbLeftPadding);
      final thumbDesiredDy =
          (size.height / 2.0) - (_sliderDecoration.thumbHeight / 2.0);
      final thumbCenter = Offset(thumbDesiredDx, thumbDesiredDy);

      canvas.drawRRect(
        RRect.fromRectAndCorners(
          thumbCenter &
              Size(_sliderDecoration.thumbWidth, _sliderDecoration.thumbHeight),
          topRight: Radius.circular(_sliderDecoration.borderRadius),
          topLeft: Radius.circular(_sliderDecoration.borderRadius),
          bottomRight: Radius.circular(_sliderDecoration.borderRadius),
          bottomLeft: Radius.circular(_sliderDecoration.borderRadius),
        ),
        thumbPainter,
      );
    }

    /// Restoring the saved canvas
    canvas.restore();
  }

  /// Helped to Use the horizontal drag gesture for the slider
  HorizontalDragGestureRecognizer _drag = HorizontalDragGestureRecognizer();

  /// Indicates that our widget handles the gestures
  @override
  bool hitTestSelf(Offset position) => true;

  /// Handles the events
  @override
  void handleEvent(PointerEvent event, BoxHitTestEntry entry) {
    assert(
      debugHandleEvent(event, entry),
      'renderer should handle the events',
    );
    if (event is PointerDownEvent) {
      _drag.addPointer(event);
    }
  }

  /// Used to update the slider thumb position
  void _updateSliderThumbPosition(
    Offset localPosition,
    ChangeValueType changeValueType,
  ) {
    /// Clamp the position between the full width of the render object
    var dx = localPosition.dx.clamp(0.0, size.width);

    /// Make the size between 0 and 1 with only 1 decimal and multiply it.
    var desiredDx = double.parse((dx / size.width).toStringAsFixed(1));
    _value = desiredDx * (_max - _min) + _min;

    switch (changeValueType) {
      case ChangeValueType.changeStart:
        _onChangeStart?.call(_value);
        break;
      case ChangeValueType.change:
        _onChanged(_value);
        break;
      case ChangeValueType.changeEnd:
        _onChangeEnd?.call(_value);
        break;
    }

    /// Calling the paint and layout method to render the slider widget
    markNeedsPaint();
    markNeedsSemanticsUpdate();
  }

  @override
  void detach() {
    /// Disposing the horizontal drag gesture
    _drag.dispose();
    super.detach();
  }
}

enum ChangeValueType {
  changeStart,
  change,
  changeEnd,
}


example/lib/main.dart
import 'dart:developer' as developer;

import 'package:flutter/material.dart';
import 'package:slider_controller/slider_controller.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Slider Controller',
      home: MyHomePage(),
      debugShowCheckedModeBanner: false,
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        color: const Color(0xFF0D1835),
        padding: const EdgeInsets.all(24.0),
        alignment: Alignment.center,
        child: SliderController(
          value: 100.0,
          onChanged: (value) {
            developer.log('Slider Value : $value');
          },
          min: 50.0,
          max: 150.0,
          sliderDecoration: SliderDecoration(
            inactiveColor: const Color(0xFF1A2942),
            activeColor: const Color(0xFF2BA1C3),
            thumbColor: const Color(0xFF1E7893),
            borderRadius: 20.0,
            height: 50.0,
            isThumbVisible: true,
            thumbHeight: 25.0,
            thumbWidth: 5.0,
          ),
        ),
      ),
    );
  }
}
